src/plugins/sensor/ROSIMUSensor/ROSIMUSensor.cpp

```diff
@@ -61,136 +62,408 @@ ROSIMUSensor::ROSIMUSensor() {}
 
 void ROSIMUSensor::init(std::map<std::string, std::string> &params) {
 
-    if (!ros::isInitialized()) {
-        int argc = 0;
-        // scrimmage handles it's own SIGINT/SIGTERM shutdown in main.cpp
-        ros::init(argc, NULL, "scrimmage", ros::init_options::NoSigintHandler);
-    }
-    nh_ = std::make_shared<ros::NodeHandle>();
+  if (!ros::isInitialized()) {
+    int argc = 0;
+    // scrimmage handles it's own SIGINT/SIGTERM shutdown in main.cpp
+    ros::init(argc, NULL, "scrimmage", ros::init_options::NoSigintHandler);
+  }
+  nh_ = std::make_shared<ros::NodeHandle>();
 
-    // Setup robot namespace
-    ros_namespace_ = sc::get<std::string>("ros_namespace_prefix", params, "robot");
-    ros_namespace_ += std::to_string(parent_->id().id());
+  // Setup robot namespace
+  ros_namespace_ = sc::get<std::string>("ros_namespace_prefix", params, "robot");
+  ros_namespace_ += std::to_string(parent_->id().id());
 
-    // Create Publisher
-    imu_pub_ = nh_->advertise<sensor_msgs::Imu>(ros_namespace_ + "/imu", 1);
+  // Create Publisher
+  imu_pub_ = nh_->advertise<sensor_msgs::Imu>(ros_namespace_ + "/imu", 1);
 
     // Open imu_data CSV for append (app) and set column headers
     std::string csv_filename = parent_->mp()->log_dir() + "/imu_data_robot" + std::to_string(parent_->id().id()) + ".csv";
     if (!csv.open_output(csv_filename, std::ios_base::app)) std::cout << "Couldn't create csv file" << endl;
     if (!csv.output_is_open()) cout << "File isn't open. Can't write to CSV" << endl;
-    csv.set_column_headers("time, dt, ENU_POSX, ENU_POSY, ENU_POSZ, ENU_VELX, ENU_VELY, ENU_VELZ, BodytoNED_roll, BodytoNED_pitch, BodytoNED_yaw");
 
-    prev_vel_ = parent_->state_truth()->vel();
-    prev_quat_ = parent_->state_truth()->quat();
-    prev_time_ = time_->t();
+  csv.set_column_headers("time, dt, ECEF_POSX, ECEF_POSY, ECEF_POSZ, ECEF_VELX, ECEF_VELY, ECEF_VELZ, bodyToEcef_X, bodyToEcef_Y, bodyToEcef_Z, bodyToEcef_W, dV_X, dV_Y, dV_Z, dTheta_X, dTheta_Y, dTheta_Z");
+
+  prev_time_ = time_->t();
+  first_sample_collected = false;
+}
+
+Eigen::Vector3d ROSIMUSensor::lla_to_ecef(double lat, double lon, double alt) {
+  Eigen::Vector3d ecef;
+
+  double factor = sin(lat);
+  factor = sqrt(1 - ecc * ecc * factor * factor);
+  ecef(0) = (earth_radius / factor + alt) * cos(lat) * cos(lon);
+  ecef(1) = (earth_radius / factor + alt) * cos(lat) * sin(lon);
+  ecef(2) = ((earth_radius * (1 - ecc * ecc)) / factor + alt) * sin(lat);
+
+  return ecef;
+}
+
+Eigen::Vector3d ROSIMUSensor::ecef_to_lla(Eigen::Vector3d ecef) {
```

Comment:
We already have GeographicLib as a dependency for scrimmage. Can you take a look at the Forward() and Reverse() functions that it provides already to compute ECEF/LLA point conversions: https://geographiclib.sourceforge.io/html/classGeographicLib_1_1Geocentric.html

Unfortunately, GeographicLib doesn't provide orientation conversions, so you will have to perform those manually. By using libraries like GeographicLib we reduce the possibility of bugs creeping into the code base.

Reply:
Sounds good, I believe I incorporated GeographicLib where appropriate in my lastest commit, in addition to the forward and reverse conversions I also derived all of my constants from it.